### DISCUSSION TITLE
Select query returns duplicated entities when one-to-many relation is used

### DISCUSSION THREAD

**Discussion by ksajdak** [mikro-orm/mikro-orm#3316] [Q&A]
<p dir="auto"><strong>Describe the bug</strong><br>

Following entities are defined:

```typescript
@Entity({tableName: 'subject'})
export class Subject {
    @PrimaryKey({primary: true, type: 'number'})
    public id!: number;

    @Property({type: 'string', length: 50})
    public name!: string;

    @OneToMany(() => Attenders, attender => attender.subject)
    public attenders = new Collection<Attenders>(this);
}

@Entity({tableName: 'attenders'})
export class Attenders {
    public [PrimaryKeyType]?: [number, number];

    @ManyToOne(() => Subject, {primary: true})
    public subject!: Subject;

    @PrimaryKey({primary: true, type: 'number'})
    public readonly studentId!: number;
}
```

Table subject:
```sql
+--+----------+
|id|name      |
+--+----------+
|1 |Maths     |
|2 |Literature|
|3 |Biology   |
+--+----------+
```
Table: attenders:
```
+----------+----------+
|subject_id|student_id|
+----------+----------+
|1         |1         |
|1         |2         |
|2         |1         |
+----------+----------+

```

And I want to retrieve subjects that student id 1 attends to, so:
```typescript
const queryBuilder = this.entityManager
                .createQueryBuilder(Subject)
                .join('attenders', 'attenders')
                .where({attenders: {student_id: 1}})
                .limit(10)
                .offset(0);

const subjects = await queryBuilder.execute();
```

Actual result: Maths, Maths, Literature
Expected result: Maths, Literature
</p>

<p dir="auto"><strong>Additional context</strong><br>
It works with mikro-orm 5.0.2.
It seems like it stopped work after this feature https://github.com/mikro-orm/mikro-orm/commit/db9963fff8ceb980354b328f2d59353b9177aef3  (Automatic PAGINATE if one-to-many relation is detected. It wraps my query in another few queries.)

sql with mirko-orm@5.0.2
```sql
SELECT `s0`.*
FROM `subject` AS `s0`
INNER JOIN `attenders` AS `attenders` ON `s0`.`id` = `attenders`.`subject_id`
WHERE `attenders`.`student_id` = 1
LIMIT 10;
```

sql with mikro-orm@5.2.3:
```sql
SELECT `s0`.*
FROM `subject` AS `s0`
INNER JOIN `attenders` AS `attenders` ON `s0`.`id` = `attenders`.`subject_id`
WHERE `s0`.`id` IN (SELECT `s0`.`id`
                    FROM (SELECT `s0`.`id`
                          FROM `subject` AS `s0`
                                   INNER JOIN `attenders` AS `attenders` ON `s0`.`id` = `attenders`.`subject_id`
                          WHERE `attenders`.`student_id` = 1
                          GROUP BY `s0`.`id`
                          LIMIT 10) AS `s0`);
 ```
</p>

<p dir="auto"><strong>Versions</strong></p>

Dependency | Version
-- | --
node | v16.13.2
typescript | ^4.5.4
mikro-orm | 5.2.3
mysql | 8.0.17

**Comment by B4nan** [comment]
Joining to many relation will always result in this - it is expected behaviour. Your query that got generated in 5.0.2 was wrong*, that commit you mentioned is fixing an issue you dont even know you had, as joining to many relation will always "produce more rows". You can't just join to many relation and apply limit, as that limit would work for the whole result set, not just for root entity (which is what you probably want).

(* or semi-wrong, depends on what that `{attenders: {student_id: 1}}` condition exactly does, maybe it pruned the result set so you, but we can't detect that anyhow - this depends on the actual data you have in the database)

Now what to do about this - either you are certain about the query (and you understand how to-many joins work and why pagination is there), and you can disable automatic pagination via `qb.setFlag(QueryFlag.DISABLE_PAGINATE)`, or you add distinct flag via `qb.setFlag(QueryFlag.DISTINCT)`. Either way, I would say you should use `qb.getResult()` instead of `qb.execute()`.

edit: Example for better understanding: imagine you have the classic author entity with 1:m collection of books. For simplicity, let's say we have exactly two authors, each of them having 5 books (so 10 books in total). Now if you query authors and you join this to-many relation, you end up with exactly this problem. You dont get 2 rows, you get 10 rows, because you joined the books table (the to-many relation), which can result is more than one connected row for an author. Adding `limit: 2` to this would result in getting only the first author, with 2 of his books. The pagination flag helps with this by doing that subquery magic.

**Comment by ksajdak** [comment]
@B4nan Thanks for the provided example. The problem you described is known to me. 


However:
> (* or semi-wrong, depends on what that {attenders: {student_id: 1}} condition exactly does, maybe it pruned the result set so you, but we can't detect that anyhow - this depends on the actual data you have in the database

The query is correct. I think you missed the fact that there is composite primary key on table attenders (subject_id, student_id). So it does not depend on actual data in database but the schema of a database. Thanks to that the query that consists `inner join` and `where` condition `{attenders: {student_id: 1}}` will always give me unique results of subject entity.

Could you look closer at the query I provided with mikro-orm@5.2.3? Why does it return results for `student_id: 2` despite the fact I filtered it out by condition `{attenders: {student_id: 1}}` ? (just replace first line ```SELECT `s0`.*` with `SELECT *` ```. )

**Reply by B4nan** [reply]
> So it does not depend on actual data in database

It _does_ depend on data, you are filtering the data. Your condition = dependency on data. I am not interested in parsing the condition and checking for this (the condition could be even just an SQL fragment, or object with aliases in keys, or a mixture, or some much more complex condition...).

No plans to change anything about the way pagination work, as I said, if you know what you are doing, just disable it for this QB instance.

> Why does it return results for student_id: 2 despite the fact I filtered it out by condition {attenders: {student_id: 1}} ? (just replace first line SELECT `s0`.*` with `SELECT *` . )

This filter is interpreted as "search root entities that has student=1 in their attenders collection. We use pagination to be able to fetch join relations, and those relations are supposed to return all items by default - your filter is applied only to the root entity with pagination, and the list of PKs of those root entities are then used for loading everything.

Like only thing I could imagine is not enabling pagination on QB level if you use `qb.execute()` instead of `qb.getResult()`, but that itself feels a bit wrong and semi-breaking, so still nothing I would change before v6.
